### PR TITLE
Add option to Filter to user instances

### PR DESCRIPTION
**Pull request by gitttt-1234**
Add option to filter only the user instances from the .slp file.

### Summary by CodeRabbit

- New Feature: Added a new option to the `LabelsReader` class that allows users to filter and include only user instances in the labels. This provides more flexibility in reading and processing labeled frames.
- Test: Introduced new test cases to verify the functionality of the `LabelsReader` class, specifically the new user instance filtering feature. This ensures the reliability and correctness of the new feature.

### CODE REVIEW

File: sleap_nn/data/providers.py [hunk 1]
```diff
@@ -14,17 +14,26 @@ class LabelsReader(IterDataPipe):
...
         """Initialize labels attribute of the class."""
         self.labels = labels
 
+        # Filter to user instances
+        if user_instances:
+            for lf in self.labels:
+                lf.instances = lf.user_instances
+            self.labels.labeled_frames = [
+                lf for lf in self.labels.labeled_frames if len(lf) > 0
+            ]
```
[the remark this answers is not in the document]
Comment (talmo) [reply sleap_nn/data/providers.py:30]: Agreed. We should also make it clear that it's a flag, like calling it `user_instances_only`

File: tests/data/test_providers.py [hunk 2]
```diff
@@ -9,3 +11,61 @@ def test_providers(minimal_instance):
...
+    l = LabelsReader(labels, user_instances=True)
+    # Check user instance filtering
+    assert len(list(l)) == 1
+
+    pred_lf = sio.LabeledFrame(video=video, frame_idx=0, instances=[pred_inst])
+    labels = sio.Labels(
+        videos=[video], skeletons=[skeleton], labeled_frames=[user_lf, pred_lf]
+    )
+    l = LabelsReader(labels, user_instances=False)
+    assert len(list(l)) == 2
```
Comment (talmo) [inline tests/data/test_providers.py:71]: Move this to a separate test and add a test for when you have a labeled frame with 1 predicted instance and no user instances.

**Review by talmo — CHANGES_REQUESTED** [review]

File: sleap_nn/data/providers.py [hunk 3]
```diff
@@ -16,17 +17,31 @@ class LabelsReader(IterDataPipe):
...
         labels: sleap_io.Labels object that contains LabeledFrames that will be
             accessed through a torchdata DataPipe
+        user_instances_only: True if filter labels only to user instances else False. Default value True
     """
 
-    def __init__(self, labels: sio.Labels) -> None:
+    def __init__(self, labels: sio.Labels, user_instances_only: bool = True):
         """Initialize labels attribute of the class."""
-        self.labels = labels
+        self.labels = copy.deepcopy(labels)
```
Comment (roomrys) [inline sleap_nn/data/providers.py:25]:
```suggestion
```

**Review by talmo — APPROVED** [review]

**Review by talmo — APPROVED** [review]
Approved conditional on reverting to deepcopy -- let's see what happens!

**Pull request merged**